### DISCUSSION TITLE
Allow extending Kademlia with custom queries

### DISCUSSION THREAD

**Discussion by nazar-pc** [libp2p/rust-libp2p#2476] [Ideas]
The core reason for #2475 is that the system I'm working on is designed such that certain values are actually implicit and don't need to be stored in Kademlia.

With that in mind, I'd prefer to benefit from available implementation otherwise, namely recursive lookup, maybe even quorum feature for those custom values.

My first hacky idea was to write a custom [RecordStore](https://docs.rs/libp2p/latest/libp2p/kad/record/store/trait.RecordStore.html) implementation that will reach out to custom logic and will refuse to store anything from Kademlia side (acting as read-only data store as far as Kademlia implementation is concerned).

Maybe there is a room for something like custom queries that is a more generalize version of currently available logic that developer can customize by providing a generic parameter (the default will be `()`, meaning not custom queries)?

**Comment by mxinden** [comment]
I am not 100% I understand what you are referring to with "queries". Are you referring to queries against the local record store, or queries to other peers on the wire, or ...?

**Reply by nazar-pc** [reply]
Queries to other peers. Existing logic that finds closest peers to issue `PUT_VALUE` to could be reused or logic for `GET_VALUE` could be extended to cover custom more cases (see https://github.com/libp2p/rust-libp2p/issues/2478).
